### DISCUSSION TITLE
Custom expression shortcuts

### DISCUSSION THREAD

**Discussion by francipvb** [sqlalchemy/sqlalchemy#8083] [Usage Questions] [answered]
Hello all,

I want to add some shortcuts to expressions on types. In a particular use case, I want to compare against year/month/day/etc. on a datetime construct without using the `extract` construct.

```python
from sqlalchemy.orm.decl_api import as_declarative
from sqlalchemy import Column, Integer, DateTime

@as_declarative
class Base:
    id=Column(Integer, primary_key=True, autoincrement=True)

class Item(Base):
    created=Column(DateTime, nullable=False)
```

Now I want to do something like this:

```python
from sqlalchemy import select

query=select(Item).where(Item.created.year==2020)
```

How do you implement this?

Thank you so much.

**Comment by zzzeek** [comment, marked as answer]
this can be achieved using custom comparators, take a look at https://docs.sqlalchemy.org/en/14/core/custom_types.html#redefining-and-creating-new-operators


here's that first Integer example adapted to what you have

```py
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import extract
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.orm import as_declarative


@as_declarative()
class Base:
    id = Column(Integer, primary_key=True, autoincrement=True)


class EnhancedDateTime(DateTime):
    class comparator_factory(DateTime.Comparator):
        @property
        def year(self):
            return extract("year", self.expr)


class Item(Base):
    __tablename__ = "i"
    id = Column(Integer, primary_key=True)
    created = Column(EnhancedDateTime, nullable=False)


query = select(Item).where(Item.created.year == 2020)

print(query)

```

```sql
SELECT i.id, i.created 
FROM i 
WHERE EXTRACT(year FROM i.created) = :param_1
```

**Reply by francipvb** [reply]
Thank you!

I skipped this from the documentation... 😳

**Reply by zzzeek** [reply]
perfectly fine question, this is a deep corner of things and I think our docs still lack good navigation.  it's when people ask about major concepts like how to commit a session or something that it gets a little tedious to point out the docs.
